# Problem Statement

For the next set of questions, we will be using census data from the United States in the United States. This dataset contains population data for countries and states in the US from 2010 to 2015. 

1. Which state has the most countries in it? (hint: consider the sumlevel key carefully! You'll need this for future questions too...)
2. Only looking at the three most populous countries for each state, what are the three most populous states (in order of highest population to lowest population)? Use CENSUS2010POP
3. Which city has the most countries in it?
4. Which region has the most division in it?

In [3]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
# read and import file
df = pd.read_csv(r"C:\Users\Ogundele Yakub\Desktop\DA Projects\DAFI Project\Python Projects\census.csv")
df

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME            CTYNAME  \
0         40       3         6      1       0  Alabama            Alabama   
1         50       3         6      1       1  Alabama     Autauga County   
2         50       3         6      1       3  Alabama     Baldwin County   
3         50       3         6      1       5  Alabama     Barbour County   
4         50       3         6      1       7  Alabama        Bibb County   
...      ...     ...       ...    ...     ...      ...                ...   
3188      50       4         8     56      37  Wyoming  Sweetwater County   
3189      50       4         8     56      39  Wyoming       Teton County   
3190      50       4         8     56      41  Wyoming       Uinta County   
3191      50       4         8     56      43  Wyoming    Washakie County   
3192      50       4         8     56      45  Wyoming      Weston County   

      CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010  ...  \
0           4779736            4780127          4785161  ...   
1             54571              54571            54660  ...   
2            182265             182265           183193  ...   
3             27457              27457            27341  ...   
4             22915              22919            22861  ...   
...             ...                ...              ...  ...   
3188          43806              43806            43593  ...   
3189          21294              21294            21297  ...   
3190          21118              21118            21102  ...   
3191           8533               8533             8545  ...   
3192           7208               7208             7181  ...   

      RDOMESTICMIG2011  RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  \
0             0.002295         -0.193196          0.381066          0.582002   
1             7.242091         -2.915927         -3.012349          2.265971   
2            14.832960         17.647293         21.845705         19.243287   
3            -4.728132         -2.500690         -7.056824         -3.904217   
4            -5.527043         -5.068871         -6.201001         -0.177537   
...                ...               ...               ...               ...   
3188          1.072643         16.243199         -5.339774        -14.252889   
3189         -1.589565          0.972695         19.525929         14.143021   
3190        -17.755986         -4.916350         -6.902954        -14.215862   
3191        -11.637475         -0.827815         -2.013502        -17.781491   
3192        -11.752361         -8.040059         12.372583          1.533635   

      RDOMESTICMIG2015  RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  \
0            -0.467369     1.030015     0.826644     1.383282     1.724718   
1            -2.530799     7.606016    -2.626146    -2.722002     2.592270   
2            17.197872    15.844176    18.559627    22.727626    20.317142   
3           -10.543299    -4.874741    -2.758113    -7.167664    -3.978583   
4             0.177258    -5.088389    -4.363636    -5.403729     0.754533   
...                ...          ...          ...          ...          ...   
3188        -14.248864     1.255221    16.243199    -5.295460   -14.075283   
3189         -0.564849     0.654527     2.408578    21.160658    16.308671   
3190        -12.127022   -18.136812    -5.536861    -7.521840   -14.740608   
3191          1.682288   -11.990126    -1.182592    -2.250385   -18.020168   
3192          6.935294   -12.032179    -8.040059    12.372583     1.533635   

      RNETMIG2015  
0        0.712594  
1       -2.187333  
2       18.293499  
3      -10.543299  
4        1.107861  
...           ...  
3188   -14.070195  
3189     1.520747  
3190   -12.606351  
3191     1.441961  
3192     6.935294  

[3193 rows x 100 columns]

# Data Inspection

In [5]:
# shape of the data
df.shape

(3193, 100)

In [6]:
# info of the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3193 entries, 0 to 3192
Data columns (total 100 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   SUMLEV                 3193 non-null   int64  
 1   REGION                 3193 non-null   int64  
 2   DIVISION               3193 non-null   int64  
 3   STATE                  3193 non-null   int64  
 4   COUNTY                 3193 non-null   int64  
 5   STNAME                 3193 non-null   object 
 6   CTYNAME                3193 non-null   object 
 7   CENSUS2010POP          3193 non-null   int64  
 8   ESTIMATESBASE2010      3193 non-null   int64  
 9   POPESTIMATE2010        3193 non-null   int64  
 10  POPESTIMATE2011        3193 non-null   int64  
 11  POPESTIMATE2012        3193 non-null   int64  
 12  POPESTIMATE2013        3193 non-null   int64  
 13  POPESTIMATE2014        3193 non-null   int64  
 14  POPESTIMATE2015        3193 non-null   int64  
 15  NPO

In [8]:
# columns of the data
df.columns.tolist()

['SUMLEV',
 'REGION',
 'DIVISION',
 'STATE',
 'COUNTY',
 'STNAME',
 'CTYNAME',
 'CENSUS2010POP',
 'ESTIMATESBASE2010',
 'POPESTIMATE2010',
 'POPESTIMATE2011',
 'POPESTIMATE2012',
 'POPESTIMATE2013',
 'POPESTIMATE2014',
 'POPESTIMATE2015',
 'NPOPCHG_2010',
 'NPOPCHG_2011',
 'NPOPCHG_2012',
 'NPOPCHG_2013',
 'NPOPCHG_2014',
 'NPOPCHG_2015',
 'BIRTHS2010',
 'BIRTHS2011',
 'BIRTHS2012',
 'BIRTHS2013',
 'BIRTHS2014',
 'BIRTHS2015',
 'DEATHS2010',
 'DEATHS2011',
 'DEATHS2012',
 'DEATHS2013',
 'DEATHS2014',
 'DEATHS2015',
 'NATURALINC2010',
 'NATURALINC2011',
 'NATURALINC2012',
 'NATURALINC2013',
 'NATURALINC2014',
 'NATURALINC2015',
 'INTERNATIONALMIG2010',
 'INTERNATIONALMIG2011',
 'INTERNATIONALMIG2012',
 'INTERNATIONALMIG2013',
 'INTERNATIONALMIG2014',
 'INTERNATIONALMIG2015',
 'DOMESTICMIG2010',
 'DOMESTICMIG2011',
 'DOMESTICMIG2012',
 'DOMESTICMIG2013',
 'DOMESTICMIG2014',
 'DOMESTICMIG2015',
 'NETMIG2010',
 'NETMIG2011',
 'NETMIG2012',
 'NETMIG2013',
 'NETMIG2014',
 'NETMIG2015',
 'RES

In [10]:
# check for missing values
df.isnull().sum()

SUMLEV         0
REGION         0
DIVISION       0
STATE          0
COUNTY         0
              ..
RNETMIG2011    0
RNETMIG2012    0
RNETMIG2013    0
RNETMIG2014    0
RNETMIG2015    0
Length: 100, dtype: int64

In [11]:
# summary statistucs of the data frame
df.describe().astype("int")

SUMLEV  REGION  DIVISION  STATE  COUNTY  CENSUS2010POP  \
count    3193    3193      3193   3193    3193           3193   
mean       49       2         5     30     101         193387   
std         1       0         1     15     107        1176201   
min        40       1         1      1       0             82   
25%        50       2         4     18      33          11299   
50%        50       3         5     29      77          26424   
75%        50       3         7     45     133          71404   
max        50       4         9     56     840       37253956   

       ESTIMATESBASE2010  POPESTIMATE2010  POPESTIMATE2011  POPESTIMATE2012  \
count               3193             3193             3193             3193   
mean              193396           193765           195251           196744   
std              1176244          1178710          1189647          1200508   
min                   82               83               90               81   
25%                11299            11275            11277            11195   
50%                26446            26467            26417            26362   
75%                71491            71721            72387            72496   
max             37254503         37334079         37700034         38056055   

       ...  RDOMESTICMIG2011  RDOMESTICMIG2012  RDOMESTICMIG2013  \
count  ...              3193              3193              3193   
mean   ...                -1                -2                -1   
std    ...                10                11                12   
min    ...              -128              -147              -109   
25%    ...                -6                -7                -6   
50%    ...                -1                -2                -1   
75%    ...                 2                 2                 3   
max    ...               101               119               239   

       RDOMESTICMIG2014  RDOMESTICMIG2015  RNETMIG2011  RNETMIG2012  \
count              3193              3193         3193         3193   
mean                 -2                -1            0           -1   
std                  11                12           10           12   
min                -178               -84         -128         -147   
25%                  -7                -6           -5           -6   
50%                  -1                -2            0           -1   
75%                   3                 3            4            3   
max                 146               281          101          119   

       RNETMIG2013  RNETMIG2014  RNETMIG2015  
count         3193         3193         3193  
mean             0           -1            0  
std             12           11           12  
min           -102         -178          -62  
25%             -5           -6           -5  
50%              0            0            0  
75%              4            4            5  
max            239          146          281  

[8 rows x 98 columns]

#### 
Based on the provided data, we have information on various demographic and population statistics for different regions, divisions, states, and counties. The data includes population figures from the 2010 census, estimates base for 2010, and population estimates for the years 2010 to 2015. Additionally, it includes information on domestic migration (RDOMESTICMIG) and net migration (RNETMIG) for the corresponding years.

The statistics show that the average population count across all regions is around 49, with a standard deviation of 1. The census2010pop ranges from a minimum of 40 to a maximum of 840. The average estimated base population in 2010 is around 193,000, with a standard deviation of approximately 1.1 million.

The migration data indicates that the average domestic migration is close to zero, with a small standard deviation. The net migration data also shows a similar pattern.

In [13]:
# confirm data type
df.dtypes

SUMLEV           int64
REGION           int64
DIVISION         int64
STATE            int64
COUNTY           int64
                ...   
RNETMIG2011    float64
RNETMIG2012    float64
RNETMIG2013    float64
RNETMIG2014    float64
RNETMIG2015    float64
Length: 100, dtype: object

In [47]:
# Top 5 states
# Assuming you have a DataFrame called 'df' with a column 'state' and 'street_name'
# Count the states
state_counts = df['STATE'].value_counts()
state_counts

# Group by state name and return the state names
grouped_states = df.groupby('STNAME').size().reset_index(name='count')
grouped_states

# Top 5 stATE
top_5_states = grouped_states.nlargest(5, 'count')
print(top_5_states)

      STNAME  count
43     Texas    255
10   Georgia    160
46  Virginia    134
17  Kentucky    121
25  Missouri    116


## Question 1
The state with the most county in it is Texas with state code number 43 and it has 255 county in it.
Georgia follow Texas with 160 county, Virginia is the third state with 134 county in it.  
Kentucky is the 4th state with 121 county.
Missouri is the 5th state with 116 county.


In [33]:
# Assuming 'df' is your DataFrame
df['CENSUS2010POP'] = pd.to_numeric(df['CENSUS2010POP'], errors='coerce')

# Group by county name and calculate the sum of CENSUS2010POP
county_sums = df.groupby('CTYNAME')['CENSUS2010POP'].sum()

# Group by state name and calculate the sum of CENSUS2010POP
state_sums = df.groupby('STNAME')['CENSUS2010POP'].sum()

# Return the top 5 county by population
top_5_county = county_sums.nlargest(3)
# Return the top 5 state by population
top_5_state = state_sums.nlargest(3)
print(top_5_county)
print(top_5_state)


CTYNAME
California    37253956
Texas         25145561
New York      19378102
Name: CENSUS2010POP, dtype: int64
STNAME
California    74507912
Texas         50291122
New York      38756204
Name: CENSUS2010POP, dtype: int64


## Question 2

The most populous county is California county with a population of over 37 million(37,253,956) follow by Texas county with population of over 25 million(25,145,561).
New York county is the third populous county with over 19 million(19,378.102).

The most populous state is California with a population of over 74 million(74,507,912) followed by Texas with a population of over 50 milllion and thirdly New York with total population of over 38 million people(38,756,204)

In [36]:
# Top 5 states
# Assuming you have a DataFrame called 'df' with a column 'state' and 'street_name'
# Count the states
state_counts = df['COUNTY'].value_counts()
state_counts

# Group by street name and return the street names
grouped_streets = df.groupby('CTYNAME').size().reset_index(name='count')
grouped_streets

# Top 5 street names
top_5_streets = grouped_streets.nlargest(5, 'count')
print(top_5_streets)


                CTYNAME  count
1822  Washington County     30
855    Jefferson County     25
609     Franklin County     24
849      Jackson County     23
1004     Lincoln County     23


## Question 3
The city with the most county is Washington with 30 county followed by Jefferson with 25 countys.
The thrid city with the most county Franklin with 24 countys and Jackson city is the fourth city with 23 countys.
Lincoln county has 23 countys in it.

In [42]:
# Distinct count of the 'DIVISION' column
distinct_division_count = df['DIVISION'].nunique()
grouped_streets = df.groupby('REGION').size().reset_index(name='count')
print(distinct_division_count)
print(grouped_streets)

9
   REGION  count
0       1    226
1       2   1067
2       3   1439
3       4    461


## Question 4
The region with the most division in it is region 3